### PR TITLE
fix: Allow the mode setup/creation to be async, and provide a few more values to extension/app config/mode setup.

### PR DESCRIPTION
**Pull request by wayfarer3130**
### Context

Adds the extension/mode loader to the app config setup so that things listed in the loader can be provided/used.

#### PR

- [] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [] OS: 
- [] Node version: 
- [] Browser:
  

[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch

### CODE REVIEW

File: platform/app/src/appInit.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ async function appInit(appConfigOrFunc, defaultExtensions, defaultModes) {
 
   const appConfig = {
     ...(typeof appConfigOrFunc === 'function'
-      ? await appConfigOrFunc({ servicesManager })
+      ? await appConfigOrFunc({ servicesManager, loadModules })
```
Comment (wayfarer3130) [inline platform/app/src/appInit.js:43]: The load modules is added here so that the app config can use modules listed in the loadModules.

File: platform/app/src/appInit.js [hunk 2]
```diff
@@ -111,7 +111,7 @@ async function appInit(appConfigOrFunc, defaultExtensions, defaultModes) {
           ? appConfig.modesConfiguration[id]
           : {};
 
-      mode = mode.modeFactory({ modeConfiguration });
+      mode = await mode.modeFactory({ modeConfiguration, loadModules });
```
Comment (wayfarer3130) [inline platform/app/src/appInit.js:114]: And the modeFactory is made async capable so that it can wait for files to load/modules internal to load.

**Comment by sedghi** [comment]
I understand your intention. I asked Ibrahim to initiate a similar effort for dev dependencies like percy, cypress, docusaurus, etc. Here is his PR: https://github.com/OHIF/Viewers/pull/4188/files

I prefer his approach. What are your thoughts, @wayfarer3130?

File: extensions/cornerstone-dynamic-volume/package.json [hunk 3]
```diff
@@ -35,6 +35,7 @@
     "@ohif/i18n": "3.9.0-beta.31",
     "@ohif/ui": "3.9.0-beta.31",
     "dcmjs": "^0.29.5",
+    "dcmjs": "*",
```
Comment (sedghi) [inline extensions/cornerstone-dynamic-volume/package.json:38]: duplicate

File: platform/cli/templates/extension/dependencies.json [hunk 4]
```diff
@@ -57,6 +57,6 @@
     "eslint-loader": "^2.0.0",
     "webpack": "^5.50.0",
     "webpack-merge": "^5.7.3",
-    "webpack-cli": "^4.7.2"
+    "webpack-cli": "^5.0.2"
```
Comment (sedghi) [inline platform/cli/templates/extension/dependencies.json:60]: i think we are using another version elsewhere

File: platform/app/.webpack/webpack.pwa.js [hunk 5]
```diff
@@ -108,6 +108,7 @@ module.exports = (env, argv) => {
             globOptions: {
               ignore: ['**/*.min.js.map'],
             },
+            noErrorOnMissing: true,
```
Comment (sedghi) [inline platform/app/.webpack/webpack.pwa.js:111]: please add documentation

File: platform/app/.webpack/writePluginImportsFile.js [hunk 6]
```diff
@@ -46,6 +45,7 @@ function getFormattedWindowBlock(addToWindowLines) {
   let content =
     'const extensions = [];\n' +
     'const modes = [];\n' +
+    'window.modules ||= {}; \n' +
```
Comment (sedghi) [inline platform/app/.webpack/writePluginImportsFile.js:48]: as discussed we will separate this into anther PR

File: package.json [hunk 7]
```diff
@@ -117,14 +123,15 @@
     "jest-canvas-mock": "^2.1.0",
     "jest-environment-jsdom": "^29.5.0",
     "jest-junit": "^6.4.0",
-    "lerna": "^7.2.0",
+    "lerna": "^5.4.2",
```
Comment (sedghi) [inline package.json:126]: undo

File: externals/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
+# External Dependencies
+
+This module contains optional dependencies for including in OHIF, such
+as the DICOM Microscopy Viewer component.  The intent is to allow both
+runtime and compiled/built external components, with the naming distinguishing
+between default externals and custom externals.
+
+* `external-<NAME>` is used for default externals.  These are included in git/core deploy
```
Comment (sedghi) [inline externals/README.md:8]: let's move to a path based similar to other routers e.g. next.js

File: externals/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
+# External Dependencies
```
Comment (sedghi) [inline externals/README.md:1]: we can call the parent folder `addOns`

File: .circleci/config.yml [hunk 10]
```diff
@@ -23,14 +23,14 @@ executors:
       - image: cimg/node:<< parameters.node-version >>-browsers
     parameters:
       node-version:
-        default: '18.16.1'
+        default: '20.14.0'
```
Comment (sedghi) [inline .circleci/config.yml:26]: we can revert this and also frozen below

File: platform/cli/templates/extension/dependencies.json (hunk 4 above)
Comment (wayfarer3130) [reply platform/cli/templates/extension/dependencies.json:60]: Updating that overall, so leaving this one.

File: package.json [hunk 11]
```diff
@@ -138,9 +144,10 @@
     "stylus": "^0.59.0",
     "stylus-loader": "^7.1.3",
     "terser-webpack-plugin": "^5.1.4",
-    "typescript": "4.6.4",
+    "typescript": "5.4.5",
```
Comment (sedghi) [inline package.json:147]: please rever this until we move cornerstone3D to 5.x too

File: package.json [hunk 12]
```diff
@@ -162,11 +169,14 @@
...
     "**/@babel/runtime": "^7.20.13",
     "commander": "8.3.0",
+    "dcmjs": ">=0.33.0",
+    "dicomweb-client": ">=0.10.4",
     "nth-check": "^2.1.1",
     "trim-newlines": "^5.0.0",
     "glob-parent": "^6.0.2",
     "trim": "^1.0.0",
     "package-json": "^8.1.0",
+    "typescript": "5.4.5",
```
Comment (sedghi) [inline package.json:179]: here too

File: platform/app/netlify.toml [hunk 13]
```diff
@@ -19,7 +19,7 @@
 [build.environment]
   # If 'production', `yarn install` does not install devDependencies
   NODE_ENV = "development"
-  NODE_VERSION = "18.16.1"
+  NODE_VERSION = "20.14.0"
```
Comment (sedghi) [inline platform/app/netlify.toml:22]: revert 18

File: modes/segmentation/src/index.tsx [hunk 14]
```diff
@@ -156,7 +156,7 @@ function modeFactory({ modeConfiguration }) {
     /** List of extensions that are used by the mode */
     extensions: extensionDependencies,
     /** HangingProtocol used by the mode */
-    // hangingProtocol: [''],
+    // hangingProtocol: ['@ohif/mnGrid'],
```
Comment (sedghi) [inline modes/segmentation/src/index.tsx:159]: I guess you wanted to include mnGrid right?

Comment (wayfarer3130) [reply modes/segmentation/src/index.tsx:159]: It doesn't really work to make it an empty string, so I wanted an example that is what we normally get by default.

Comment (wayfarer3130) [reply modes/segmentation/src/index.tsx:159]: Added a comment

File: extensions/default/src/DicomWebDataSource/index.js [hunk 15]
```diff
@@ -393,10 +393,10 @@ function createDicomWebApi(dicomWebConfig, servicesManager) {
           // The value.Value will be set with the bulkdata read value
           // in which case it isn't necessary to re-read this.
           if (value && value.BulkDataURI && !value.Value) {
+            // handle the scenarios where bulkDataURI is relative path
+            fixBulkDataURI(value, naturalized, dicomWebConfig);
```
Comment (wayfarer3130) [inline extensions/default/src/DicomWebDataSource/index.js:397]: This changes fixes pdf and video display for certain external URLs.  Probably should have been aprt of a different PR, so if you want me to move it, I can.

File: extensions/default/src/DicomWebDataSource/utils/fixBulkDataURI.ts [hunk 16]
```diff
@@ -20,16 +20,23 @@ function fixBulkDataURI(value, instance, dicomWebConfig) {
   // in case of the relative path, make it absolute. The current DICOM standard says
   // the bulkdataURI is relative to the series. However, there are situations where
   // it can be relative to the study too
-  if (!value.BulkDataURI.startsWith('http') && !value.BulkDataURI.startsWith('/')) {
-    if (dicomWebConfig.bulkDataURI?.relativeResolution === 'studies') {
-      value.BulkDataURI = `${dicomWebConfig.wadoRoot}/studies/${instance.StudyInstanceUID}/${value.BulkDataURI}`;
+  const { BulkDataURI } = value;
```
Comment (wayfarer3130) [inline extensions/default/src/DicomWebDataSource/utils/fixBulkDataURI.ts:23]: This fixes some issues display PDF/video from external data sources (eg not default).  Could be part of another PR

File: extensions/default/src/getHangingProtocolModule.js [hunk 17]
```diff
@@ -42,6 +42,7 @@ const defaultProtocol = {
         // Try to match series with images by default, to prevent weird display
         // on SEG/SR containing studies
         {
+          weight: 10,
```
Comment (wayfarer3130) [inline extensions/default/src/getHangingProtocolModule.js:45]: This is fixing an integration test that had gotten disabled.

File: extensions/dicom-pdf/.webpack/webpack.prod.js [hunk 18]
```diff
@@ -42,6 +44,10 @@ module.exports = (env, argv) => {
       new webpack.optimize.LimitChunkCountPlugin({
         maxChunks: 1,
       }),
+      new MiniCssExtractPlugin({
```
Comment (wayfarer3130) [inline extensions/dicom-pdf/.webpack/webpack.prod.js:47]: Fixes a build issue that suddenly appeared - not sure why we started needing this.

File: platform/app/public/config/default.js [hunk 19]
```diff
@@ -62,9 +62,10 @@ const config = {
         omitQuotationForMultipartRequest: true,
       },
     },
+
     {
       namespace: '@ohif/extension-default.dataSourcesModule.dicomweb',
-      sourceName: 'dicomweb2',
+      sourceName: 'ohif2',
```
Comment (wayfarer3130) [inline platform/app/public/config/default.js:68]: Adding a few endpoints for testing.

**Comment by wayfarer3130** [comment]
Running the minimal version is:
yarn install:dev -f
yarn test:e2e:serve

then to switch back:
yarn install -f
yarn test:e2e:serve

**Review by sedghi — APPROVED** [review]
Thanks, it looks great i tested it
